File: others/valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> a;
        
        for (char c : s) {
            switch (c) {
                case ('('):
                    a.push_back(')');
                    break;
                case ('{'):
                    a.push_back('}');
                    break;
                case ('['):
                    a.push_back(']');
                    break;
                case (')'):
                    if (a.size() == 0) return false;
                    else if (c != a.back()) return false;
                    a.pop_back();
                    break;
                case ('}'):
                    if (a.size() == 0) return false;
                    else if (c != a.back()) return false;
                    a.pop_back();
                    break;
                case (']'):
                    if (a.size() == 0) return false;
                    else if (c != a.back()) return false;
                    a.pop_back();
                    break;
            }
        }
        
        return a.size() == 0;
    }
};

File: easy/others/Valid_Parentheses.java
class Solution {
    // by using stack to solve
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        char[] parentheses = s.toCharArray();
        for (int i = 0; i < parentheses.length; i++) {
            // traverse the whole char array;
            if (parentheses[i] == '(' || parentheses[i] == '[' || parentheses[i] == '{') {
                // push to stack if opening parentheses;
                stack.push(parentheses[i]);
            } else {
                if (stack.empty()) {
                    // edge case: start by closing ones;
                    return false;
                }
                char to_compare = stack.peek();
                if (to_compare == '(' && parentheses[i] == ')' || to_compare == '[' && parentheses[i] == ']' || to_compare == '{' && parentheses[i] == '}') {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }
        return (stack.empty());

    }
}